Fix for Issue#17 - Supporting custom schemas in incremental models

rahulj51 opened the pull request.

findinpath commented:
@rahulj51 thank you for submitting the PR.

Could you please reword your commit to 

```
Support custom schemas in incremental models

Resolves: #17 
```

Can you also add a test for your PR?

rahulj51 commented:
Fixed the commit message. The tests may take a while for me to write as I am not familiar with the `dbt_adpater_tests` plugin. I'll give it a shot.

findinpath commented:
We have an incremental model already here: https://github.com/starburstdata/dbt-trino/blob/master/docker/dbt/jaffle_shop/models/marts/core/dim_order_dates.sql

We may create a new schema in the memory connector or simply define a new memory connector here 

https://github.com/starburstdata/dbt-trino/tree/master/docker/trino/catalog

Then with
```
make dbt-trino-tests
```

you should be able to test whether your new change actually works 


Explaining this in writing makes it clear to me that the tests in place need to be improved

rahulj51 commented:
I am sure I am missing something but how does the Makefile run the models and tests inside jaffle_shop? When I run 
`make dbt-trino-tests`, it only triggers the unit and integration tests, none of which run the models inside `jaffle_shop`.

Shouldn't the Makefile internally call `run_tests.bash` ?

mdesmet commented:
@rahulj51: i also saw that, I think it's missing that call.

rahulj51 commented:
> @rahulj51: i also saw that, I think it's missing that call.

I added the call to the Makefile. And changed the `jaffle-shop` incremental model to use a custom schema in order to test the fix. 

Let me know if this makes sense.

findinpath commented:
@rahulj51 

sorry for misleading you


```
./docker/dbt/build.sh
./docker/init_trino.bash
./docker/run_tests.bash
```

This should be what you need.

As described here:
https://github.com/starburstdata/dbt-trino#running-tests

findinpath commented on docker/dbt/jaffle_shop/models/marts/core/dim_order_dates.sql:
```diff
@@ -1,5 +1,5 @@
 {{
-    config(materialized = 'incremental')
+    config(materialized = 'incremental', schema='custom_schema')
```
please do create a different model for showcasing this functionality

rahulj51 replied:
Will do.

findinpath commented on Makefile:
```diff
@@ -3,6 +3,7 @@ dbt-trino-tests:
 	./docker/dbt/build.sh
 	./docker/init_trino.bash
 	pip install -r dev_requirements.txt
+	./docker/run_tests.bash
```
Leave the `Makefile` out of this PR because fixing it is not really in the scope.

rahulj51 replied:
ok. Will revert this.

hovaesco replied:
oh I will add them in separate PR, thanks for catching this

findinpath commented on dbt/include/trino/macros/materializations/incremental.sql:
```diff
@@ -3,7 +3,7 @@
     {%- set dest_cols_csv = dest_columns | map(attribute='quoted') | join(', ') -%}
     insert into {{ target_relation }}
-    select {{dest_cols_csv}} from {{ tmp_relation.include(database=false, schema=false) }};
+    select {{dest_cols_csv}} from {{ tmp_relation.include(database=true, schema=true) }};
```
i see the following statement on Trino while executing the jaffle shop model tests

```
insert into "memory"."default_custom_schema"."dim_order_dates"
    select "order_date" from "memory"."default_custom_schema"."dim_order_dates__dbt_tmp"
```

Also
```
trino> show schemas in memory;
        Schema         
-----------------------
 default               
 default_custom_schema 
 information_schema    
(3 rows)
```

Creating new schemas is not the purpose of this PR right?

rahulj51 replied:
No, the purpose is to support custom schemas if they exist. The incremental model fails if a custom schema is used.

findinpath replied:
I had a look again at how the tests are running.

dbt is issuing this command: 

```
create schema if not exists "memory"."default_custom_schema"
```

although the incremental model is defined to run on "custom_schema"

```
{{
    config(materialized = 'incremental', schema='custom_schema')
}}
```

Please take a look again at your implementation.

BTW, pls take a look at 
https://trino.io/docs/current/installation/cli.html#examples

to setup the creation of the "custom_schema" in the Trino docker instance before starting dbt.

This can be done by simply adding a new bash script which does

```
docker container exec -t dbt-trino-trino-1  trino --execute "create schema memory.custom_schema"
```

findinpath replied:
Creating the schema in case it doesn't exist can be done via dbt. 
However it is important that the created schema does fit the specified name.

rahulj51 replied:
This is due to how dbt treats custom schemas - https://docs.getdbt.com/docs/building-a-dbt-project/building-models/using-custom-schemas. It basically concatenates the `<target_schema>_<custom_schema>`  by default. 

In other words, if I define a custom schema called `custom_schema` , dbt would internally translate it to `<default_schema>_<custom_schema>`

findinpath replied:
Thanks for the heads'up @rahulj51  . 

Could you please add a README entry  about https://docs.getdbt.com/docs/building-a-dbt-project/building-models/using-custom-schemas ?
> ## Custom schemas
> By default, all dbt models are built in the schema specified in your target. In dbt projects with lots of models, it may be useful to instead build some models in schemas other than your target schema – this can help logically group models together.
> See [Using custom schemas](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/using-custom-schemas) for more details.

rahulj51 replied:
@hovaesco will do by tomorrow.

rahulj51 replied:
@hovaesco , I added the section on custom schemas to the README

rahulj51 commented:
1. Added a new incremental model for custom schema.
2. Removed the changes from Makefile
3. Modified `run_tests.sh` to rerun the incremental model

hovaesco reviewed (approved):
Looks good, thank you for the contribution! 
@findinpath PTAL
